ffi/custom_passes.cpp

```diff
@@ -0,0 +1,613 @@
+        for (Instruction &ii : *bb) {
+            if (isRelatedDecref(incref, &ii) != NULL) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    template <class T>
+    bool eraseNullFirstArgFromList(T& refops) {
+        bool mutated = false;
+        for (CallInst*& refop: refops) {
+            if (!isNonNullFirstArg(refop)) {
+                refop->eraseFromParent();
+                mutated |= true;
+                refop = NULL;
+            }
+        }
+        return mutated;
+    }
+
+    /**
+     * Find related decrefs to incref inside a basicblock in order
+     */
+    std::vector<CallInst*> findRelatedDecrefs(BasicBlock* bb, CallInst* incref) {
+        std::vector<CallInst*> res;
+        for (Instruction &ii : *bb) {
+        CallInst *call_inst;
+        if ((call_inst = isRelatedDecref(incref, &ii))){
+            res.push_back(call_inst);
+        } else {
+            continue;
+        }
+        }
+        return res;
+    }
+
+    CallInst* isRelatedDecref(CallInst *incref, Instruction *ii) {
+        // TODO: DRY
+        if (ii->getOpcode() == Instruction::Call) {
+            CallInst *call_inst = dyn_cast<CallInst>(ii);
+            Value *callee = call_inst->getCalledOperand();
+            if ( callee->getName() != "NRT_decref" ) {
+                return NULL;
+            }
+            if (incref->getArgOperand(0) != call_inst->getArgOperand(0)) {
+                return NULL;
+            }
+            return call_inst;
+        }
+        return NULL;
+    }
+
+    void getAnalysisUsage(AnalysisUsage &Info) const override {
+        Info.addRequired<DominatorTreeWrapperPass>();
+        Info.addRequired<PostDominatorTreeWrapperPass>();
+    }
+
+    bool isNonNullFirstArg(CallInst *call_inst){
+        auto val = call_inst->getArgOperand(0);
+        auto ptr = dyn_cast<ConstantPointerNull>(val);
+        return ptr == NULL;
+    }
+
+    bool hasAnyDecrefInNode(BasicBlock *bb) {
+
+        for (Instruction &ii: *bb) {
+            CallInst* refop = GetRefOpCall(&ii);
+            if (refop != NULL && IsDecRef(refop)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Pre-condition: head_node dominates tail_node
+     */
+    bool hasDecrefBetweenGraph(BasicBlock *head_node, BasicBlock *tail_node,
+                               SmallVector<BasicBlock*, 20> &stack) {
+        if (basicBlockInList(head_node, stack)) {
+            return false;
+        }
+        if (DEBUG_PRINT) {
+            errs() << "Check..." << head_node->getName() << "\n";
+        }
+
+        if (hasAnyDecrefInNode(head_node)) return true;
+
+        stack.push_back(head_node);
+        Instruction *term = head_node->getTerminator();
+        for (unsigned i=0; i < term->getNumSuccessors(); ++i) {
+            BasicBlock *child = term->getSuccessor(i);
+            if (child == tail_node)
+                return false;
```

Comment:
Is there a risk of this returning false too early? It seems that the first time the tail node is hit, if it's down a path that there's no decref, then false will be returned - could this get hit and return false whilst there is a decref still to be found down another not-yet-visited path?

Reply:
You are right!